Adding support for third-party Controllers (e.g. OpenShift)

baderbuddy opened the pull request:
When pulling ReplicationController it will now look for an OwnerReference and roll up duplicates to that owner. This allows Polaris to work with OpenShift DeploymentConfigs but also any other CRD.

Also added a quick check for dividing by 0.

rbren commented on pkg/validator/summary.go:
```diff
@@ -17,6 +17,9 @@ type CountSummaryByCategory map[string]CountSummary
 // GetScore returns an overall score in [0, 100] for the CountSummary
 func (cs CountSummary) GetScore() uint {
 	total := (cs.Successes * 2) + cs.Warnings + (cs.Errors * 2)
+	if total == 0 {
```
Thanks! I can't believe [this bug is still open](https://github.com/FairwindsOps/polaris/issues/206) 😬

baderbuddy commented on pkg/validator/controller.go:
```diff
@@ -32,24 +33,60 @@ func ValidateController(conf *conf.Configuration, controller controller.Interfac
 	var controllersToAudit []controller.Interface
 	for _, supportedControllers := range config.ControllersToScan {
-		loadedControllers, _ := controllers.LoadControllersByKind(supportedControllers, kubeResources)
+		loadedControllers, err := controllers.LoadControllersByKind(supportedControllers, kubeResources)
```
It looks like Jobs and CronJobs were throwing some kind of error, I'd hate to return the error and kill everything based on what could be a mistake in the config. So I think logging makes sense.

baderbuddy commented on pkg/validator/controller.go:
```diff
@@ -32,24 +33,60 @@ func ValidateController(conf *conf.Configuration, controller controller.Interfac
+	// If an owner exists then set the name to the controller.
+	// This allows us to handle CRDs creating Controllers or DeploymentConfigs in OpenShift.
+	if len(owners) > 0 {
+		firstOwner := owners[0]
```
None, I hope it's not possible but the types seem to technically allow it.

baderbuddy commented:
What was the original thinking with the ControllersToScan option? That might help me in thinking through how to fit this in. I was trying to think through any kind of scenario where I wouldn't just want to scan every kind of controller that I could.

rbren commented:
TBH, I don't fully remember 😅 - the guy who implemented it is no longer on the project. I wouldn't be opposed to removing it for v1.

I still have some concerns over the current implementation though - e.g. I think we'll end up seeing multiple `CronJob` objects - one for the original, and one for each `Job` it's created. And if the ownership is XController -> Pod, with no intermediate controller type, we'll miss it.

I do dislike our current pattern of needing a bunch of boilerplate code for each controller type, but I think it'd be good to stay consistent between vanilla k8s and open shift. Maybe we can sync offline and come up with a better path forward. I think checking pod parents like you're doing here might be an interesting route.

baderbuddy commented:
Robert, do you want to take another look? I think I have the dynamic querying for parents working, I even have some dynamicness working for loading from a YAML file. I think this is at least the right track, but there's probably more to clean up.

rbren commented on pkg/validator/controllers/interface.go:
```diff
@@ -13,16 +11,33 @@ import (
 type Interface interface {
```
Do we still need an interface, if `GenericController` is the only implementer?

rbren commented on pkg/kube/resources.go:
```diff
@@ -54,18 +50,12 @@ func CreateResourceProvider(directory string) (*ResourceProvider, error) {
+		SourceName:    directory,
+		Nodes:         []corev1.Node{},
+		Namespaces:    []corev1.Namespace{},
+		Pods:          []corev1.Pod{},
```
Rather than `Pods`, maybe it makes sense for this to be `Controllers: []GenericController{}`. Then we can do all the Owner resolution here, instead of inside `Validate`.

If possible, I'd like to keep interactions w/ the k8s API here, rather than passing around an interface.

rbren commented on pkg/kube/resources.go:
```diff
@@ -168,222 +137,77 @@ func CreateResourceProviderFromAPI(kube kubernetes.Interface, clusterName string
+		finalDoc["apiVersion"] = "v1"
+		finalDoc["kind"] = "Pod"
+		finalDoc["spec"] = getPodSpec(yamlNode)
+		marshelledYaml, err := yaml.Marshal(finalDoc)
```
`s/mareshell/marshal` (below too)

rbren commented on pkg/kube/resources.go:
```diff
@@ -168,222 +137,77 @@ func CreateResourceProviderFromAPI(kube kubernetes.Interface, clusterName string
+		finalDoc := make(map[string]interface{})
+		finalDoc["metadata"] = yamlNode["metadata"]
+		finalDoc["apiVersion"] = "v1"
+		finalDoc["kind"] = "Pod"
```
I think we're losing the `Kind` from the file here. E.g. if the user has a `Deployment` manifest in their repo, we want that to show up as a `Deployment` in the dashboard.

rbren commented on pkg/validator/fullaudit_test.go:
```diff
@@ -33,8 +33,8 @@ func TestGetTemplateData(t *testing.T) {
 		Successes: uint(0),
-		Warnings:  uint(9),
-		Errors:    uint(9),
+		Warnings:  uint(1),
```
Looks like we lost a lot of tests - why is that?

baderbuddy commented on pkg/config/supportedcontrollers.go:
```diff
@@ -10,7 +10,7 @@ import (
 	appsv1beta2 "k8s.io/api/apps/v1beta2"
 	batchv1 "k8s.io/api/batch/v1"
-	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	batchV1beta1 "k8s.io/api/batch/v1beta1"
```
This was actually a mistake, I don't know how I accidentally deleted this at some point and was trying to put it back in, I'll revert it back to lowercase.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/FairwindsOps/polaris?pullRequest=240) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/FairwindsOps/polaris?pullRequest=240) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: baderbuddy<br/>:x: Robert Brennan<br/><hr/>**Robert Brennan** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/FairwindsOps/polaris?pullRequest=240) it.</sub>

baderbuddy commented:
I fixed up all of the tests, a few don't have as much data and I added TODOs there, that's because we're only retrieving things at the pod level now and all of the fake controllers haven't added pods.

rbren commented on pkg/validator/controllers/generic.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+// LoadControllers loads a list of controllers from the kubeResources Pods
+func LoadControllers(pods []kubeAPICoreV1.Pod, dynamicClientPointer *dynamic.Interface, restMapperPointer *meta.RESTMapper) []GenericController {
```
Can we move this logic into `pkg/kube`? I'd like to keep all the API interactions in there.

rbren commented on pkg/validator/controllers/generic.go:
```diff
@@ -0,0 +1,126 @@
+	}
+	// If an owner exists then set the name to the controller.
+	// This allows us to handle CRDs creating Controllers or DeploymentConfigs in OpenShift.
+	for len(owners) > 0 {
```
Should we log an info message if there are multiple owners?

baderbuddy replied:
Got the logging in there now.

rbren commented on pkg/validator/controllers/generic.go:
```diff
@@ -0,0 +1,126 @@
+	PodSpec     kubeAPICoreV1.PodSpec
+	ObjectMeta  kubeAPIMetaV1.ObjectMeta
+	Kind        config.SupportedController
+	KindString  string
```
I don't like the idea of tracking both `Kind` and `KindString` - it's not immediately apparent why they'd diverge (though I can see from the code that they do)

Is there anything wrong with always setting `Kind` to the controller Kind, and getting rid of the `SupportedController` type?

baderbuddy replied:
I got rid of the Kind on the GenericController, but left the SupportedController type because it still seemed useful for a few things, especially it seems like we use it on the validation webhook side.

rbren commented on pkg/validator/controllers/cronjob.go:
```diff
@@ -2,42 +2,20 @@ package controllers
-	controller.Name = originalDeploymentResource.Name
-	controller.Namespace = originalDeploymentResource.Namespace
-	controller.K8SResource = originalDeploymentResource
+func NewCronJobController(originalResource kubeAPIBatchV1beta1.CronJob) GenericController {
```
Seems like these functions are just used in the tests - possible to nuke them or move them to the `/test` package?

baderbuddy replied:
The validator webhook still uses these too

rbren commented on pkg/validator/output.go:
```diff
@@ -62,11 +59,12 @@ type ResultSet map[string]ResultMessage
+	Kind        string
+	Results     ResultSet
+	PodResult   PodResult
+	CreatedTime time.Time
```
Why did this get added?

baderbuddy replied:
The time was there specifically for the dedupe logic
